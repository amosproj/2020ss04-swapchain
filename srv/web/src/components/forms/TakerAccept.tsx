import React, { useState, useEffect } from "react"
import { fakeAddress } from "../../util"
import { Label } from "./Label"
import { Input } from "./Input"
import { SubmitButton } from "./SubmitButton"
import ACCS, { ACCSConfig } from "../../accs/accs"
import { ECPair } from "bitcoinjs-lib"
import { getSecret } from "../../pkg/secret/secret"

export const TakerAccept = () => {
    const [fields, setFields] = useState({
        bitcoinPrivateKey: "",
        bitsharesPrivateKey: "",
        bitcoinPublicAddress: "", 
        bitsharesAccountName: "",
        hash: getSecret(),
        amount: "",
        timelockDuration: 6, 
    })

    const [isValid, setValid] = useState(false)
    const [showError, setShowError] = useState(false)


    //validation
    useEffect(() => {
        const isValid = fields.bitcoinPrivateKey !== "" &&
            fields.bitsharesPrivateKey !== "" &&
            fields.hash !== ""


        setValid(isValid)
        if (isValid) {
            setShowError(false)
        }
    }, [fields])

    const updateField = (e: React.FormEvent<HTMLInputElement>) => {
        setFields({
            ...fields,
            [e.currentTarget.name]: e.currentTarget.value
        })
    }

    const submitHandler = () => {
        if (!isValid) {
            setShowError(true)
            return
        }

        const accs = new ACCS(fields.bitcoinPrivateKey)
        const config = {} as ACCSConfig
        //config.txType = fields. -  does this even belong to the ProposerAccept? 
        config.privateKeyBTS = fields.bitsharesPrivateKey
        //config.keyPairCompressedBTC = fields.bitcoinPublicAddress
        config.accountBTS = fields.bitsharesAccountName
        config.amount = fields.amount // still put amout field in UI 
        config.timelockBTC = fields.timelockDuration
        config.secret= fields.hash // we dont have to generate a secret hash again - but secret does not accept a string, only getSecret()
        console.log(fields)
    }
    return (
        <form className="container mx-auto">
            <section className="mt-12 ">
                <div className="-mx-3 md:flex">
                    <div className="px-3 md:w-1/2">
                        <Label label="Bitcoin private key"></Label>
                        <Input name="bitcoinPrivateKey" value={fields.bitcoinPrivateKey} onChange={updateField} type="text" placeholder={fakeAddress(15)}></Input>
                    </div>
                    <div className="px-3 mt-4 md:w-1/2 md:mt-0">
                        <Label label="Bitshares private key"></Label>
                        <Input name="bitsharesPrivateKey" value={fields.bitsharesPrivateKey} onChange={updateField} type="text" placeholder={fakeAddress(15)}></Input>
                    </div>
                </div>
                <p className="mt-2 text-sm text-center text-gray-500 ">
                    Your private keys will never leave your browser, it is only used to sign your transactions. <a href="/" className="relative text-xs text-blue-500">Read more in our docs.</a>
                </p>
            </section>

            <section className="mt-12 -mx-3 md:flex">

                <div className="px-3 md:w-1/2">
                    <Label label="Bitcoin public address"></Label>
                    <Input name="bitcoinPublicAddress" value={fields.bitcoinPublicAddress} onChange={updateField} type="text" placeholder={fakeAddress(30)}></Input>
                </div>
                <div className="px-3 mt-4 md:w-1/2 md:mt-0">
                    <Label label="Bitshares account name"></Label>
                    <Input name="bitsharesAccountName" value={fields.bitsharesAccountName} onChange={updateField} type="text" placeholder="amos"></Input>
                </div>
            </section>
            
            <section className="mt-12">

                <div className="block">
                    <Label label="Hash from Counterparty"></Label>
                    <Input name="hash" value={fields.hash} onChange={updateField} type="text" placeholder="Please enter the hash you received from your trading partner"></Input>
                </div>

            </section>

            <section className="mt-12">          
                <div className="md:w-1/2">
                        <Label label="Timelock duration (1 block is roughly 10 minutes)"></Label>
                        <div className="relative">
                            <span className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600">
                                Blocks          </span>
                            <Input name="timelockDuration" value={fields.timelockDuration} onChange={updateField} type="number" placeholder="6"></Input>
                        </div>
                        <p className="mt-2 text-sm text-center text-gray-500 ">
                        As a taker always make sure to set a timelock shorter than the proposer.</p>
                    </div>
                </section>




            <footer className="pt-12 mt-8 text-center border-t border-gray-300 f">
                <p className={`-mt-4 pb-4 text-red-500 text-sm ${showError ? "" : "hidden"}`}>Please fill in all fields</p>
                <SubmitButton color={isValid ? "teal" : "gray"} label="Submit" onClick={submitHandler}></SubmitButton>


            </footer>
        </form>
    )
}