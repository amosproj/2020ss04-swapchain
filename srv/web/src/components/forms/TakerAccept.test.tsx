import React from 'react';
import {TakerAccept} from './TakerAccept';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer
    .create(<TakerAccept></TakerAccept>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});