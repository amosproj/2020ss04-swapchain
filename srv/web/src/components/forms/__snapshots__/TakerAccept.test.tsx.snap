// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="container mx-auto"
>
  <section
    className="mt-12 "
  >
    <div
      className="-mx-3 md:flex"
    >
      <div
        className="px-3 md:w-1/2"
      >
        <label
          className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
        >
          Bitcoin private key
        </label>
        <input
          className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
          name="bitcoinPrivateKey"
          onChange={[Function]}
          placeholder="9229410371f0b0a..."
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="px-3 mt-4 md:w-1/2 md:mt-0"
      >
        <label
          className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
        >
          Bitshares private key
        </label>
        <input
          className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
          name="bitsharesPrivateKey"
          onChange={[Function]}
          placeholder="amos"
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <p
      className="mt-2 text-sm text-center text-gray-500 "
    >
      Your private keys will never leave your browser, it is only used to sign your transactions. 
      <a
        className="relative text-xs text-blue-500"
        href="/"
      >
        Read more in our docs.
      </a>
    </p>
  </section>
  <section
    className="mt-12"
  >
    <div
      className="block"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        Hash from Counterparty
      </label>
      <input
        className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
        name="hash"
        onChange={[Function]}
        placeholder="Please enter the hash you received from your trading partner"
        required={true}
        type="text"
        value=""
      />
    </div>
  </section>
  <footer
    className="pt-12 mt-8 text-center border-t border-gray-300 f"
  >
    <p
      className="-mt-4 pb-4 text-red-500 text-sm hidden"
    >
      Please fill in all fields
    </p>
    <button
      className="bg-gray-100 border-gray-400 inline-flex px-8 py-3 font-semibold text-gray-900 uppercase transition duration-200 ease-in-out border focus:outline-none rounded-sm hover:bg-white"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </footer>
</form>
`;
